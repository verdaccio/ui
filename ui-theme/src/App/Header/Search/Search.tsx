import debounce from 'lodash/debounce';
import React, { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { RouteComponentProps, withRouter } from 'react-router';
import AutoComplete from 'verdaccio-ui/components/AutoComplete';

import { SearchResultWeb } from '@verdaccio/types';

import { Dispatch, RootState } from '../../../store/store';

const CONSTANTS = {
  API_DELAY: 300,
  ABORT_ERROR: 'AbortError',
};

const Search: React.FC<RouteComponentProps> = ({ history }) => {
  const { t } = useTranslation();
  const { suggestions } = useSelector((state: RootState) => state.search);
  const isLoading = useSelector((state: RootState) => state?.loading?.models.search);
  const dispatch = useDispatch<Dispatch>();
  /**
   * Cancel all the requests which are in pending state.
   */
  const cancelAllSearchRequests = useCallback(() => {
    dispatch.search.clearRequestQueue();
    dispatch.search.saveSearch({ suggestions: [] });
  }, [dispatch]);

  /**
   * As user focuses out from input, we cancel all the request from requestList
   * and set the API state parameters to default boolean values.
   */
  const handleOnBlur = useCallback(
    (event: React.SyntheticEvent) => {
      // stops event bubbling
      event.stopPropagation();
      cancelAllSearchRequests();
    },
    [cancelAllSearchRequests]
  );

  /**
   * When an user select any package by clicking or pressing return key.
   */
  const handleClickSearch = useCallback(
    (event: React.SyntheticEvent, value: SearchResultWeb, reason: string): void => {
      // stops event bubbling
      event.stopPropagation();
      switch (reason) {
        case 'selectOption':
          history.push(`/-/web/detail/${value.name}`);
          break;
      }
    },
    [history]
  );

  /**
   * Fetch packages from API.
   * For AbortController see: https://developer.mozilla.org/en-US/docs/Web/API/AbortController
   */
  const handleFetchPackages = useCallback(
    ({ value }: { value: string }) => {
      if (value?.trim() !== '') {
        dispatch.search.getSuggestions({ value });
      }
    },
    [dispatch]
  );

  return (
    <AutoComplete
      onCleanSuggestions={handleOnBlur}
      onSelectItem={handleClickSearch}
      onSuggestionsFetch={debounce(handleFetchPackages, CONSTANTS.API_DELAY)}
      placeholder={t('search.packages')}
      suggestions={suggestions}
      suggestionsLoading={isLoading}
    />
  );
};

export default withRouter(Search);
