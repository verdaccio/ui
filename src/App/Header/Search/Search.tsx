import debounce from 'lodash/debounce';
import React, { useState, FormEvent, useCallback } from 'react';
import { SuggestionSelectedEventData } from 'react-autosuggest';
import { useTranslation } from 'react-i18next';
import { RouteComponentProps, withRouter } from 'react-router';

import AutoComplete from 'verdaccio-ui/components/AutoComplete';
import { callSearch } from 'verdaccio-ui/utils/calls';

import SearchAdornment from './SearchAdornment';

const CONSTANTS = {
  API_DELAY: 300,
  ABORT_ERROR: 'AbortError',
};

const Search: React.FC<RouteComponentProps> = ({ history }) => {
  const { t } = useTranslation();
  const [suggestions, setSuggestions] = useState([]);
  const [loaded, setLoaded] = useState(false);
  const [search, setSearch] = useState('');
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);
  const [requestList, setRequestList] = useState<Array<{ abort: () => void }>>([]);

  /**
   * Cancel all the requests which are in pending state.
   */
  const cancelAllSearchRequests = useCallback(() => {
    requestList.forEach(request => request.abort());
    setRequestList([]);
  }, [requestList, setRequestList]);

  /**
   * As user focuses out from input, we cancel all the request from requestList
   * and set the API state parameters to default boolean values.
   */
  const handleOnBlur = useCallback(
    (event: FormEvent<HTMLInputElement>) => {
      // stops event bubbling
      event.stopPropagation();
      setLoaded(false);
      setLoading(false);
      setError(false);
      cancelAllSearchRequests();
    },
    [setLoaded, setLoading, cancelAllSearchRequests, setError]
  );

  /**
   * onChange method for the input element.
   */
  const handleSearch = useCallback(
    (event: FormEvent<HTMLInputElement>, { newValue, method }) => {
      // stops event bubbling
      event.stopPropagation();
      if (method === 'type') {
        const value = newValue.trim();

        setLoading(true);
        setError(false);
        setSearch(value);
        setLoaded(false);
        /**
         * A use case where User keeps adding and removing value in input field,
         * so we cancel all the existing requests when input is empty.
         */
        if (value.length === 0) {
          cancelAllSearchRequests();
        }
      }
    },
    [cancelAllSearchRequests]
  );

  /**
   * Cancel all the request from list and make request list empty.
   */
  const handlePackagesClearRequested = useCallback(() => {
    setSuggestions([]);
  }, [setSuggestions]);

  /**
   * When an user select any package by clicking or pressing return key.
   */
  const handleClickSearch = useCallback(
    (
      event: FormEvent<HTMLInputElement>,
      { suggestionValue, method }: SuggestionSelectedEventData<unknown>
    ): void | undefined => {
      // stops event bubbling
      event.stopPropagation();
      switch (method) {
        case 'click':
        case 'enter':
          setSearch('');
          history.push(`/-/web/detail/${suggestionValue}`);
          break;
      }
    },
    [history]
  );

  /**
   * Fetch packages from API.
   * For AbortController see: https://developer.mozilla.org/en-US/docs/Web/API/AbortController
   */
  const handleFetchPackages = useCallback(
    async ({ value }: { value: string }) => {
      try {
        const controller = new window.AbortController();
        const signal = controller.signal;
        // Keep track of search requests.
        setRequestList([...requestList, controller]);
        const suggestions = await callSearch(value, signal);
        // @ts-ignore FIXME: Argument of type 'unknown' is not assignable to parameter of type 'SetStateAction<never[]>'
        setSuggestions(suggestions);
        setLoaded(true);
      } catch (error) {
        /**
         * AbortError is not the API error.
         * It means browser has cancelled the API request.
         */
        if (error.name === CONSTANTS.ABORT_ERROR) {
          setError(false);
          setLoaded(false);
        } else {
          setError(true);
          setLoaded(false);
        }
      } finally {
        setLoading(false);
      }
    },
    [requestList, setRequestList, setSuggestions, setLoaded, setError, setLoading]
  );

  return (
    <AutoComplete
      onBlur={handleOnBlur}
      onChange={handleSearch}
      onCleanSuggestions={handlePackagesClearRequested}
      onClick={handleClickSearch}
      onSuggestionsFetch={debounce(handleFetchPackages, CONSTANTS.API_DELAY)}
      placeholder={t('search.packages')}
      startAdornment={<SearchAdornment />}
      suggestions={suggestions}
      suggestionsError={error}
      suggestionsLoaded={loaded}
      suggestionsLoading={loading}
      value={search}
    />
  );
};

export default withRouter(Search);
