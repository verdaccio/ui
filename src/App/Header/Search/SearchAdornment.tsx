import styled from '@emotion/styled';
import Search from '@material-ui/icons/Search';
import React from 'react';

import InputAdornment from 'verdaccio-ui/components/InputAdornment';
import { Theme } from 'verdaccio-ui/design-tokens/theme';

const StyledInputAdornment = styled(InputAdornment)<{ theme?: Theme }>(props => ({
  color: props.theme && props.theme.palette.white,
}));

const SearchAdornment: React.FC = () => (
  <StyledInputAdornment position={'start'}>
    <Search />
  </StyledInputAdornment>
);

export default SearchAdornment;
