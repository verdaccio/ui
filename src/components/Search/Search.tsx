import React from 'react';
import debounce from 'lodash/debounce';
import { RouteComponentProps, withRouter } from 'react-router';
import { SuggestionSelectedEventData } from 'react-autosuggest';

import AutoComplete from '../AutoComplete';
import { callSearch } from '../../utils/calls';

import SearchAdornment from './SearchAdornment';

const CONSTANTS = {
  API_DELAY: 300,
  PLACEHOLDER_TEXT: 'Search Packages',
  ABORT_ERROR: 'AbortError',
};

const Search: React.FC<RouteComponentProps> = ({ history }) => {
  const [suggestions, setSuggestions] = React.useState([]);
  const [loaded, setLoaded] = React.useState(false);
  const [search, setSearch] = React.useState('');
  const [error, setError] = React.useState(false);
  const [loading, setLoading] = React.useState(false);
  const [requestList, setRequestList] = React.useState<Array<{ abort: () => void }>>([]);

  /**
   * Cancel all the requests which are in pending state.
   */
  const cancelAllSearchRequests = () => {
    requestList.forEach(request => request.abort());
    setRequestList([]);
  };

  /**
   * As user focuses out from input, we cancel all the request from requestList
   * and set the API state parameters to default boolean values.
   */
  const handleOnBlur = (event: React.FormEvent<HTMLInputElement>) => {
    // stops event bubbling
    event.stopPropagation();
    setLoaded(false);
    setLoading(false);
    setError(false);
    cancelAllSearchRequests();
  };

  /**
   * onChange method for the input element.
   */
  const handleSearch = (event: React.FormEvent<HTMLInputElement>, { newValue, method }) => {
    // stops event bubbling
    event.stopPropagation();
    if (method === 'type') {
      const value = newValue.trim();

      setLoading(true);
      setError(false);
      setSearch(value);
      setLoaded(false);
      /**
       * A use case where User keeps adding and removing value in input field,
       * so we cancel all the existing requests when input is empty.
       */
      if (value.length === 0) {
        cancelAllSearchRequests();
      }
    }
  };

  /**
   * Cancel all the request from list and make request list empty.
   */
  const handlePackagesClearRequested = () => {
    setSuggestions([]);
  };

  /**
   * When an user select any package by clicking or pressing return key.
   */
  const handleClickSearch = (
    event: React.FormEvent<HTMLInputElement>,
    { suggestionValue, method }: SuggestionSelectedEventData<unknown>
  ): void | undefined => {
    // stops event bubbling
    event.stopPropagation();
    switch (method) {
      case 'click':
      case 'enter':
        setSearch('');
        history.push(`/-/web/detail/${suggestionValue}`);
        break;
    }
  };

  /**
   * Fetch packages from API.
   * For AbortController see: https://developer.mozilla.org/en-US/docs/Web/API/AbortController
   */
  const handleFetchPackages = async ({ value }: { value: string }) => {
    try {
      const controller = new window.AbortController();
      const signal = controller.signal;
      // Keep track of search requests.
      setRequestList([...requestList, controller]);
      const suggestions = await callSearch(value, signal);
      // @ts-ignore FIXME: Argument of type 'unknown' is not assignable to parameter of type 'SetStateAction<never[]>'
      setSuggestions(suggestions);
      setLoaded(true);
    } catch (error) {
      /**
       * AbortError is not the API error.
       * It means browser has cancelled the API request.
       */
      if (error.name === CONSTANTS.ABORT_ERROR) {
        setError(false);
        setLoaded(false);
      } else {
        setError(true);
        setLoaded(false);
      }
    } finally {
      setLoading(false);
    }
  };

  return (
    <AutoComplete
      onBlur={handleOnBlur}
      onChange={handleSearch}
      onCleanSuggestions={handlePackagesClearRequested}
      onClick={handleClickSearch}
      onSuggestionsFetch={debounce(handleFetchPackages, CONSTANTS.API_DELAY)}
      placeholder={CONSTANTS.PLACEHOLDER_TEXT}
      startAdornment={<SearchAdornment />}
      suggestions={suggestions}
      suggestionsError={error}
      suggestionsLoaded={loaded}
      suggestionsLoading={loading}
      value={search}
    />
  );
};

export default withRouter(Search);
