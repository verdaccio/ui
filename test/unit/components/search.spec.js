/**
 * @flow
 * @prettier
 */

import React from 'react';
import { mount } from 'enzyme';

import { Search } from '../../../src/webui/components/Search/index';

const SEARCH_FILE_PATH = '../../../src/webui/components/Search/index';
const API_FILE_PATH = '../../../src/webui/utils/api';
const URL_FILE_PATH = '../../../src/webui/utils/url';

// Global mocks
const event = {
  stopPropagation: jest.fn(),
};
window.location.assign = jest.fn();

describe('<Search /> component test', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Search />);
  });

  test('should load the component in default state', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('onBlur: should cancel all search requests', async () => {
    const { onBlur, requestList } = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'cancelAllSearchRequests');

    const request = {
      abort: jest.fn(),
    };
    // adds a request for AbortController
    wrapper.instance().requestList = [request];

    onBlur(event);

    expect(request.abort).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalled();
    expect(wrapper.state('error')).toBeFalsy();
    expect(wrapper.state('loaded')).toBeFalsy();
    expect(wrapper.state('loading')).toBeFalsy();
    expect(spy).toHaveBeenCalled();
    expect(requestList).toEqual([]);
  });

  test('handleSearch: when user type package name in search component and set loading to true', () => {
    const { handleSearch } = wrapper.instance();
    const newValue = 'verdaccio';

    handleSearch(event, { newValue, method: 'type' });

    expect(event.stopPropagation).toHaveBeenCalled();
    expect(wrapper.state('error')).toBeFalsy();
    expect(wrapper.state('loaded')).toBeFalsy();
    expect(wrapper.state('loading')).toBeTruthy();
    expect(wrapper.state('search')).toEqual(newValue);
  });

  test('handleSearch: cancel all search requests when there is no value in search component with type method', () => {
    const { handleSearch, requestList } = wrapper.instance();
    const spy = jest.spyOn(wrapper.instance(), 'cancelAllSearchRequests');
    const newValue = '';

    handleSearch(event, { newValue, method: 'type' });

    expect(event.stopPropagation).toHaveBeenCalled();
    expect(wrapper.state('error')).toBeFalsy();
    expect(wrapper.state('loaded')).toBeFalsy();
    expect(wrapper.state('loading')).toBeTruthy();
    expect(wrapper.state('search')).toEqual(newValue);
    expect(spy).toHaveBeenCalled();
    expect(requestList).toEqual([]);
  });

  test('handleSearch: when method is not type method', () => {
    const { handleSearch } = wrapper.instance();
    const newValue = '';

    handleSearch(event, { newValue, method: 'click' });

    expect(event.stopPropagation).toHaveBeenCalled();
    expect(wrapper.state('error')).toBeFalsy();
    expect(wrapper.state('loaded')).toBeFalsy();
    expect(wrapper.state('loading')).toBeFalsy();
    expect(wrapper.state('search')).toEqual(newValue);
  });

  test('handlePackagesClearRequested: should clear suggestions', () => {
    const { handlePackagesClearRequested } = wrapper.instance();

    handlePackagesClearRequested();

    expect(wrapper.state('suggestions')).toEqual([]);
  });
});

describe('<Search /> component: mocks specific tests ', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.doMock('lodash/debounce', () => {
      return function debounceMock(fn, delay) {
        return fn;
      };
    });
  });

  test('handleFetchPackages: should load the packages from API', async () => {
    const apiResponse = [{ name: 'verdaccio' }, { name: 'verdaccio-htpasswd' }];
    const suggestions = [{ name: 'verdaccio' }, { name: 'verdaccio-htpasswd' }];

    jest.doMock(API_FILE_PATH, () => ({
      request(url) {
        expect(url).toEqual('search/verdaccio');
        return Promise.resolve(apiResponse);
      },
    }));

    const Search = require(SEARCH_FILE_PATH).Search;
    const component = mount(<Search />);
    component.setState({ search: 'verdaccio' });
    const { handleFetchPackages } = component.instance();
    await handleFetchPackages({ value: 'verdaccio' });

    expect(component.state('suggestions')).toEqual(suggestions);
    expect(component.state('error')).toBeFalsy();
    expect(component.state('loaded')).toBeTruthy();
    expect(component.state('loading')).toBeFalsy();
  });

  test('handleFetchPackages: when browser cancel a request', async () => {
    const apiResponse = { name: 'AbortError' };

    jest.doMock(API_FILE_PATH, () => ({ request: jest.fn(() => Promise.reject(apiResponse)) }));

    const Search = require(SEARCH_FILE_PATH).Search;
    const component = mount(<Search />);
    component.setState({ search: 'verdaccio' });
    const { handleFetchPackages } = component.instance();
    await handleFetchPackages({ value: 'verdaccio' });

    expect(component.state('error')).toBeFalsy();
    expect(component.state('loaded')).toBeFalsy();
    expect(component.state('loading')).toBeFalsy();
  });

  test('handleFetchPackages: when API server failed request', async () => {
    const apiResponse = { name: 'BAD_REQUEST' };

    jest.doMock(API_FILE_PATH, () => ({
      request(url) {
        expect(url).toEqual('search/verdaccio');
        return Promise.reject(apiResponse);
      },
    }));

    const Search = require(SEARCH_FILE_PATH).Search;
    const component = mount(<Search />);
    component.setState({ search: 'verdaccio' });
    const { handleFetchPackages } = component.instance();
    await handleFetchPackages({ value: 'verdaccio' });

    expect(component.state('error')).toBeTruthy();
    expect(component.state('loaded')).toBeFalsy();
    expect(component.state('loading')).toBeFalsy();
  });

  test('handleClickSearch: should change the window location on click or return key', () => {
    const getDetailPageURL = jest.fn(() => 'detail/page/url');
    jest.doMock(URL_FILE_PATH, () => ({ getDetailPageURL }));

    const suggestionValue = [];
    const Search = require(SEARCH_FILE_PATH).Search;
    const pushHandler = jest.fn();
    const component = mount(<Search history={{ push: pushHandler }} />);
    const { handleClickSearch } = component.instance();

    // click
    handleClickSearch(event, { suggestionValue, method: 'click' });
    expect(event.stopPropagation).toHaveBeenCalled();
    expect(pushHandler).toHaveBeenCalledTimes(1);

    // return key
    handleClickSearch(event, { suggestionValue, method: 'enter' });
    expect(event.stopPropagation).toHaveBeenCalled();
    expect(pushHandler).toHaveBeenCalledTimes(2);
  });
});
