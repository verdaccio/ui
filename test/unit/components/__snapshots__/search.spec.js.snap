// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> component test should load the component in default state 1`] = `"<div class=\\"css-1crzyyo e18292mt0\\"><div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"react-autowhatever-1\\" aria-expanded=\\"false\\" class=\\"react-autosuggest__container\\"><div class=\\"MuiFormControl-root-1 MuiFormControl-fullWidth-4 react-autosuggest__input\\" aria-autocomplete=\\"list\\" aria-controls=\\"react-autowhatever-1\\"><div class=\\"MuiInputBase-root-18 MuiInput-root-5 css-n9ojyg MuiInput-underline-9 MuiInputBase-fullWidth-27 MuiInput-fullWidth-12 MuiInputBase-formControl-19 MuiInput-formControl-6 MuiInputBase-adornedStart-22\\"><div class=\\"MuiInputAdornment-root-35 MuiInputAdornment-positionStart-37\\" style=\\"color: rgb(255, 255, 255);\\"><svg class=\\"MuiSvgIcon-root-40\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></div><input aria-invalid=\\"false\\" autocomplete=\\"off\\" class=\\"MuiInputBase-input-28 MuiInput-input-13 css-hodoyq MuiInputBase-inputAdornedStart-33\\" placeholder=\\"Search Packages\\" type=\\"text\\" value=\\"\\"></div></div><div class=\\"MuiPaper-root-49 MuiPaper-elevation2-53 react-autosuggest__suggestions-container css-1aijpo3 e18292mt1\\" id=\\"react-autowhatever-1\\" role=\\"listbox\\"></div></div></div>"`;
